core/src/smartcontracts/isi/mod.rs

```diff
@@ -225,6 +228,14 @@ pub mod error {
             }
         }
     }
+
+    /// Expression error, which occurs during expression evaluation
+    #[derive(Debug, Clone, Copy, Error, PartialEq, Eq)]
+    pub enum ExpressionError {
+        /// Encountered incompatible type of arguments
+        #[error("Expression does not support provided combination of arguments ({0}, {1})")]
+        BinaryOpIncompatibleNumericValueTypes(NumericValue, NumericValue),
```

Comment:
Is this semantically different from `MathError::DomainViolation`?

Reply:
From what i found, this error originates from `fixnum` crate:
```
    /// When someone tries to use operand out of the set of departure of the function.
    /// E.g.: when you try to compute the square root of a negative number.
```
So, it's kinda similar, but not entirely. 
Also i want to provide useful feedback for user and `DomainViolation` does not help with that.

Reply:
Re: feedback for the user - totally makes sense, I didn't catch it has to conform to what (none) information `fixnum` provides.
I would still say we should move this error into `MathError`. Although it could be argued that it's not strictly speaking a math error, `MathError` is what we return in all other cases of expressions failing, so IMO that would be less confusing than having another enum dedicated solely to this type of error.